File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        
        def helper(curr: ListNode) -> ListNode:
            if not curr.next:
                return curr
            
            nhead = helper(curr.next)
            curr.next.next = curr
            curr.next = None
            
            return nhead
        
        return helper(head)

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode nhead;
    
    public ListNode reverseList(ListNode head) {
        if(head == null){
            return head;
        }
        
        this.nhead = null;
        this.helper(head);
        return this.nhead;
    }
    
    private ListNode helper(ListNode curr){
        if(curr.next == null){
            this.nhead = curr;
            return curr;
        }
        
        this.helper(curr.next).next = curr;
        curr.next = null;
        return curr;
    }
}